src/MICore/Debugger.cs

```diff
@@ -482,10 +504,17 @@ public Task CmdStart()
             return CmdAsync("-exec-run", ResultClass.running);
         }
 
-        protected bool _requestingRealAsyncBreak = false;
-        public Task CmdBreak()
+        public enum BreakRequest
         {
-            _requestingRealAsyncBreak = true;
+            None,
+            Async,
+            Stop,
+            Internal
+        }
+        protected BreakRequest _requestingRealAsyncBreak = BreakRequest.None;
+        public Task CmdBreak(BreakRequest request)
+        {
+            _requestingRealAsyncBreak = request;
```

Comment:
If an internal break comes in when we already have an outstanding non-internal break, we don't want to blow away the field.

Suggestion: reorder BreakRequest from least important->most important and then you can do 'Max' of the current value and the new one.